File: php/Encoder.php
<?php
# php/Encoder.php
namespace bX\Scon;

use bX\TreeHash;

class Encoder {

    const COMMA = ',';
    const TAB = "\t";
    const PIPE = '|';
    const DOUBLE_QUOTE = '"';
    const COLON = ':';
    const OPEN_BRACKET = '[';
    const CLOSE_BRACKET = ']';
    const OPEN_BRACE = '{';
    const CLOSE_BRACE = '}';
    const LIST_ITEM_PREFIX = '- ';
    const NULL_LITERAL = 'null';
    const TRUE_LITERAL = 'true';
    const FALSE_LITERAL = 'false';
    const SEMICOLON = ';';
    const HEADER = '#!scon/1.0';

    private int $indent;
    private string $delimiter;
    private string $mode; # warn (default), loose, strict
    private ?string $enforce;
    private SchemaRegistry $registry;
    private array $extractedSchemas; # auto-detected schemas for dedup
    private bool $autoExtract;
    private bool $header;
    private array $warnings = [];

    public function __construct(array $options = []) {
        $this->indent = $options['indent'] ?? 1;
        $this->delimiter = $options['delimiter'] ?? self::COMMA;
        $this->mode = $options['mode'] ?? 'warn';
        $this->enforce = $options['enforce'] ?? null;
        $this->autoExtract = $options['autoExtract'] ?? false;
        $this->header = $options['header'] ?? false; # header off by default
        $this->registry = new SchemaRegistry();
        $this->extractedSchemas = [];
        $this->warnings = [];
    }

    public function getRegistry(): SchemaRegistry {
        return $this->registry;
    }

    public function getWarnings(): array {
        return $this->warnings;
    }

    # Encode PHP data to SCON string
    public function encode(mixed $data, array $schemas = [], array $responses = [], array $security = []): string {
        # Register explicit schemas
        foreach ($schemas as $name => $def) {
            $this->registry->register('s', $name, $def);
        }
        foreach ($responses as $name => $def) {
            $this->registry->register('r', $name, $def);
        }
        foreach ($security as $name => $def) {
            $this->registry->register('sec', $name, $def);
        }

        # Auto-extract repeated schemas if enabled
        if ($this->autoExtract && is_array($data)) {
            $this->detectRepeatedSchemas($data);
        }

        $lines = [];

        # Header (optional, off by default)
        if ($this->header) {
            $lines[] = self::HEADER;
        }

        # Directives
        if ($this->mode !== 'warn') {
            $lines[] = "@@{$this->mode}";
        }
        if ($this->enforce !== null) {
            $lines[] = "@@enforce({$this->enforce})";
        }

        # Schema definitions
        $allSchemas = $this->registry->getAll('s');
        if (!empty($allSchemas)) {
            if (!empty($lines)) $lines[] = '';
            foreach ($allSchemas as $name => $def) {
                $lines[] = "s:$name " . $this->encodeInline($def);
            }
        }

        # Response group definitions
        $allResponses = $this->registry->getAll('r');
        if (!empty($allResponses)) {
            if (!empty($lines)) $lines[] = '';
            foreach ($allResponses as $name => $def) {
                $lines[] = "r:$name " . $this->encodeResponseGroup($def);
            }
        }

        # Security group definitions
        $allSecurity = $this->registry->getAll('sec');
        if (!empty($allSecurity)) {
            if (!empty($lines)) $lines[] = '';
            foreach ($allSecurity as $name => $def) {
                $lines[] = "sec:$name " . $this->encodeInline($def);
            }
        }

        # Body
        if (!empty($allSchemas) || !empty($allResponses) || !empty($allSecurity)) {
            $lines[] = '';
        }

        # Explicit {} for empty object — soporta (object)[] y stdClass
        if ($this->isEmptyObject($data)) {
            $lines[] = '{}';
        } else {
            # stdClass con datos → convertir a array para iterar
            if ($data instanceof \stdClass) {
                $data = (array)$data;
            }
            foreach ($this->encodeValue($data, 0) as $line) {
                $lines[] = $line;
            }
        }

        # Prune orphan schemas: defined but never referenced in the body
        if ($this->autoExtract && !empty($allSchemas)) {
            $body = implode("\n", $lines);
            $pruned = [];
            foreach ($lines as $line) {
                if (preg_match('/^s:(\S+)\s/', $line, $m)) {
                    if (strpos($body, '@s:' . $m[1]) === false) {
                        continue; # orphan — skip
                    }
                }
                $pruned[] = $line;
            }
            $lines = $pruned;
        }

        return implode("\n", $lines);
    }

    # Encode a response group definition
    private function encodeResponseGroup(array $group): string {
        $parts = [];
        foreach ($group as $code => $def) {
            $desc = $def['description'] ?? '';
            $schemaRef = $def['schemaRef'] ?? null;
            $part = "$code:" . $this->encodeString($desc);
            if ($schemaRef !== null) {
                $part .= " @s:$schemaRef";
                if (!empty($def['overrides'])) {
                    $part .= ' ' . $this->encodeInline($def['overrides']);
                }
            }
            $parts[] = $part;
        }
        return self::OPEN_BRACE . implode(', ', $parts) . self::CLOSE_BRACE;
    }

    # Encode data as inline single-line notation
    private function encodeInline(mixed $data): string {
        if ($this->isPrimitive($data)) {
            return $this->encodePrimitive($data);
        }

        if (is_array($data) && $this->isSequentialArray($data)) {
            $items = array_map(fn($v) => $this->encodeInline($v), $data);
            return self::OPEN_BRACKET . implode(', ', $items) . self::CLOSE_BRACKET;
        }

        if (is_array($data)) {
            $parts = [];
            foreach ($data as $key => $val) {
                $encodedKey = $this->encodeKey($key);
                $encodedVal = $this->encodeInline($val);
                $parts[] = "$encodedKey:$encodedVal";
            }
            return self::OPEN_BRACE . implode(', ', $parts) . self::CLOSE_BRACE;
        }

        return '';
    }

    # Encode value with indentation (TOON-style body)
    private function encodeValue(mixed $value, int $depth): \Generator {
        if ($this->isPrimitive($value)) {
            $encoded = $this->encodePrimitive($value);
            if ($encoded !== '') {
                yield $encoded;
            }
            return;
        }

        # stdClass → treat as associative array
        if ($value instanceof \stdClass) {
            $value = (array)$value;
        }

        if (is_array($value)) {
            if ($this->isSequentialArray($value)) {
                yield from $this->encodeArray(null, $value, $depth);
            } else {
                yield from $this->encodeObject($value, $depth);
            }
        }
    }

    private function encodeObject(array $obj, int $depth): \Generator {
        foreach ($obj as $key => $val) {
            # Nested empty object: (object)[] or stdClass → key: {}
            if ($this->isEmptyObject($val)) {
                yield $this->indentedLine($depth, $this->encodeKey($key) . ': {}');
                continue;
            }
            # Normalize nested stdClass to array
            if ($val instanceof \stdClass) {
                $val = (array)$val;
            }
            if ($this->isPrimitive($val)) {
                $encodedKey = $this->encodeKey($key);
                $encodedVal = $this->encodePrimitive($val);
                yield $this->indentedLine($depth, "$encodedKey: $encodedVal");
            } elseif (is_array($val) && $this->isSequentialArray($val)) {
                yield from $this->encodeArray($key, $val, $depth);
            } elseif (is_array($val)) {
                $encodedKey = $this->encodeKey($key);
                # Check if value matches a registered schema
                $schemaRef = $this->findMatchingSchema($val);
                if ($schemaRef !== null) {
                    yield $this->indentedLine($depth, "$encodedKey: @s:$schemaRef");
                } else {
                    yield $this->indentedLine($depth, "$encodedKey:");
                    if (!empty($val)) {
                        yield from $this->encodeObject($val, $depth + 1);
                    }
                }
            }
        }
    }

    private function encodeArray(?string $key, array $array, int $depth): \Generator {
        $length = count($array);

        if ($length === 0) {
            if ($key !== null) {
                yield $this->indentedLine($depth, $this->encodeKey($key) . ': []');
            } else {
                yield $this->indentedLine($depth, '[]');
            }
            return;
        }

        # Array de primitivos
        if ($this->isArrayOfPrimitives($array)) {
            $header = $this->formatHeader($length, $key);
            $values = $this->encodeAndJoinPrimitives($array);
            yield $this->indentedLine($depth, "$header $values");
            return;
        }

        # Array de objetos (tabular)
        if ($this->isArrayOfObjects($array)) {
            $fields = $this->extractTabularHeader($array);
            if ($fields !== null) {
                yield from $this->encodeTabularArray($key, $array, $fields, $depth);
                return;
            }
        }

        # Array mixto / expandido
        yield from $this->encodeMixedArray($key, $array, $depth);
    }

    private function encodeTabularArray(?string $key, array $rows, array $fields, int $depth): \Generator {
        $length = count($rows);
        $header = $this->formatHeader($length, $key, $fields);
        yield $this->indentedLine($depth, $header);

        foreach ($rows as $row) {
            $values = [];
            foreach ($fields as $field) {
                $values[] = $row[$field] ?? null;
            }
            yield $this->indentedLine($depth + 1, $this->encodeAndJoinPrimitives($values));
        }
    }

    private function encodeMixedArray(?string $key, array $items, int $depth): \Generator {
        $header = $this->formatHeader(count($items), $key);
        yield $this->indentedLine($depth, $header);

        foreach ($items as $item) {
            if ($this->isPrimitive($item)) {
                yield $this->indentedListItem($depth + 1, $this->encodePrimitive($item));
            } elseif (is_array($item) && !$this->isSequentialArray($item)) {
                $schemaRef = $this->findMatchingSchema($item);
                if ($schemaRef !== null) {
                    yield $this->indentedListItem($depth + 1, "@s:$schemaRef");
                } else {
                    yield from $this->encodeObjectAsListItem($item, $depth + 1);
                }
            } elseif (is_array($item) && $this->isSequentialArray($item)) {
                if (empty($item)) {
                    yield $this->indentedListItem($depth + 1, '[]');
                } elseif ($this->isArrayOfPrimitives($item)) {
                    $subHeader = $this->formatHeader(count($item), null);
                    $values = $this->encodeAndJoinPrimitives($item);
                    yield $this->indentedListItem($depth + 1, "$subHeader $values");
                }
            }
        }
    }

    private function encodeObjectAsListItem(array $obj, int $depth): \Generator {
        if (empty($obj)) {
            yield $this->indentedLine($depth, self::LIST_ITEM_PREFIX);
            return;
        }

        $keys = array_keys($obj);
        $firstKey = $keys[0];
        $firstVal = $obj[$firstKey];
        $rest = array_slice($obj, 1, null, true);

        $encodedKey = $this->encodeKey($firstKey);

        if ($this->isPrimitive($firstVal)) {
            yield $this->indentedListItem($depth, "$encodedKey: " . $this->encodePrimitive($firstVal));
        } elseif (is_array($firstVal) && $this->isSequentialArray($firstVal) && empty($firstVal)) {
            yield $this->indentedListItem($depth, "$encodedKey: []");
        } elseif (is_array($firstVal) && $this->isSequentialArray($firstVal) && $this->isArrayOfPrimitives($firstVal)) {
            $hdr = $this->formatHeader(count($firstVal), null);
            $vals = $this->encodeAndJoinPrimitives($firstVal);
            yield $this->indentedListItem($depth, "$encodedKey$hdr $vals");
        } elseif (is_array($firstVal)) {
            yield $this->indentedListItem($depth, "$encodedKey:");
            yield from $this->encodeObject($firstVal, $depth + 2);
        }

        if (!empty($rest)) {
            yield from $this->encodeObject($rest, $depth + 1);
        }
    }

    # Find a registered schema that matches this data exactly
    private function findMatchingSchema(array $data): ?string {
        foreach ($this->registry->getAll('s') as $name => $def) {
            if ($data === $def) {
                return $name;
            }
        }
        return null;
    }

    # Detect repeated sub-structures and extract as schemas via TreeHash
    private function detectRepeatedSchemas(array $data): void {
        $result = TreeHash::hashTree($data, '', 2, false);

        foreach ($result['index'] as $entry) {
            if ($entry['count'] >= 2) {
                $name = $this->generateSchemaName($entry['path']);
                $this->registry->register('s', $name, $entry['data']);
            }
        }
    }

    private function generateSchemaName(string $path): string {
        $parts = explode('.', trim($path, '.'));
        # Strip list indices ([$0], [$1]) from the end — not meaningful names
        while (!empty($parts) && preg_match('/^\[\d+\]$/', end($parts))) {
            array_pop($parts);
        }
        $last = end($parts) ?: '';
        # Clean up common path segments
        $last = str_replace(['properties', 'content', 'application/json', 'schema'], '', $last);
        $last = trim($last, '.');
        return $last ?: 'auto_' . substr(hash('xxh128', $path), 0, 6);
    }

    # --- Formatting helpers ---

    private function formatHeader(int $length, ?string $key = null, ?array $fields = null): string {
        $header = '';
        if ($key !== null) {
            $header .= $this->encodeKey($key);
        }

        $delimSuffix = ($this->delimiter !== self::COMMA) ? $this->delimiter : '';
        $header .= self::OPEN_BRACKET . $length . $delimSuffix . self::CLOSE_BRACKET;

        if ($fields !== null) {
            $qFields = array_map([$this, 'encodeKey'], $fields);
            $header .= self::OPEN_BRACE . implode($this->delimiter, $qFields) . self::CLOSE_BRACE;
        }

        $header .= self::COLON;
        return $header;
    }

    private function extractTabularHeader(array $array): ?array {
        if (empty($array)) return null;

        $first = $array[0];
        if (!is_array($first) || $this->isSequentialArray($first)) return null;

        $firstKeys = array_keys($first);
        if (empty($firstKeys)) return null;

        foreach ($array as $row) {
            if (!is_array($row)) return null;
            $rowKeys = array_keys($row);
            if (count($rowKeys) !== count($firstKeys)) return null;
            foreach ($firstKeys as $fk) {
                if (!array_key_exists($fk, $row)) return null;
                if (!$this->isPrimitive($row[$fk])) return null;
            }
        }

        return $firstKeys;
    }

    # --- Primitive encoding (same as TOON) ---

    private function encodePrimitive(mixed $value): string {
        if ($value === null) return self::NULL_LITERAL;
        if (is_bool($value)) return $value ? self::TRUE_LITERAL : self::FALSE_LITERAL;
        if (is_int($value) || is_float($value)) {
            if ($value === 0 && is_float($value) && 1 / $value === -INF) return '0';
            return strval($value);
        }
        if (is_string($value)) return $this->encodeString($value);
        return '';
    }

    private function encodeString(string $value): string {
        if ($this->isSafeUnquoted($value)) return $value;
        # Warn on strings that look like incomplete structures
        if (isset($value[0]) && ($value[0] === '{' || $value[0] === '[')) {
            $this->warnings[] = "String starts with '{$value[0]}', possible incomplete structure: " . substr($value, 0, 60);
        }
        return self::DOUBLE_QUOTE . $this->escapeString($value) . self::DOUBLE_QUOTE;
    }

    private function encodeKey(string $key): string {
        if ($this->isValidUnquotedKey($key)) return $key;
        return self::DOUBLE_QUOTE . $this->escapeString($key) . self::DOUBLE_QUOTE;
    }

    private function escapeString(string $str): string {
        $escaped = str_replace('\\', '\\\\', $str);
        $escaped = str_replace('"', '\\"', $escaped);
        $escaped = str_replace("\n", '\\n', $escaped);
        $escaped = str_replace("\r", '\\r', $escaped);
        $escaped = str_replace("\t", '\\t', $escaped);
        $escaped = str_replace(';', '\\;', $escaped);
        return $escaped;
    }

    private function isSafeUnquoted(string $value): bool {
        if ($value === '') return false;
        if (in_array($value, [self::TRUE_LITERAL, self::FALSE_LITERAL, self::NULL_LITERAL], true)) return false;
        if (is_numeric($value)) return false;
        if (strpos($value, $this->delimiter) !== false) return false;
        if (preg_match('/[\s:"\\\\;@#\{\[\]\}]/', $value)) return false;
        return true;
    }

    private function isValidUnquotedKey(string $key): bool {
        if ($key === '') return false;
        if ($key[0] === '#') return false; # starts-with-# looks like comment
        if (preg_match('/[:\[\]{}"\\\\\s;@#,]/', $key)) return false;
        return true;
    }

    private function encodeAndJoinPrimitives(array $values): string {
        return implode($this->delimiter . ' ', array_map([$this, 'encodePrimitive'], $values));
    }

    private function isPrimitive(mixed $value): bool {
        return is_null($value) || is_bool($value) || is_int($value) || is_float($value) || is_string($value);
    }

    private function isSequentialArray(mixed $arr): bool {
        if (!is_array($arr) || empty($arr)) return is_array($arr);
        return array_keys($arr) === range(0, count($arr) - 1);
    }

    private function isArrayOfPrimitives(array $arr): bool {
        if (!$this->isSequentialArray($arr)) return false;
        foreach ($arr as $item) {
            if (!$this->isPrimitive($item)) return false;
        }
        return true;
    }

    private function isArrayOfObjects(array $arr): bool {
        if (!$this->isSequentialArray($arr)) return false;
        foreach ($arr as $item) {
            if (!is_array($item) || $this->isSequentialArray($item)) return false;
        }
        return true;
    }

    private function indentedLine(int $depth, string $content): string {
        return str_repeat(' ', $this->indent * $depth) . $content;
    }

    private function indentedListItem(int $depth, string $content): string {
        return $this->indentedLine($depth, self::LIST_ITEM_PREFIX . $content);
    }

    # Detect empty object: stdClass or (object)[] cast
    private function isEmptyObject(mixed $val): bool {
        if ($val instanceof \stdClass) {
            return empty((array)$val);
        }
        return false;
    }

}

File: src/encoder.rs
// scon/src/encoder.rs
// SCON Encoder — Value → SCON string

use crate::value::Value;
use indexmap::IndexMap;

pub struct Encoder {
    indent: usize,
    delimiter: char,
}

impl Encoder {
    pub fn new() -> Self {
        Self { indent: 1, delimiter: ',' }
    }

    pub fn with_indent(mut self, indent: usize) -> Self {
        self.indent = indent.max(1);
        self
    }

    pub fn with_delimiter(mut self, delimiter: char) -> Self {
        self.delimiter = delimiter;
        self
    }

    pub fn encode(&self, data: &Value) -> String {
        let mut buf = String::with_capacity(1024);
        match data {
            Value::Object(obj) if obj.is_empty() => buf.push_str("{}"),
            Value::Array(arr) if arr.is_empty() => buf.push_str("[]"),
            _ => self.encode_value(data, 0, &mut buf),
        }
        buf
    }

    fn encode_value(&self, value: &Value, depth: usize, buf: &mut String) {
        match value {
            Value::Object(obj) => self.encode_object(obj, depth, buf),
            Value::Array(arr) => self.encode_array_value(None, arr, depth, buf),
            _ => {
                self.write_primitive(value, buf);
            }
        }
    }

    fn encode_object(&self, obj: &IndexMap<String, Value>, depth: usize, buf: &mut String) {
        let mut first = true;
        for (key, val) in obj {
            if !first { buf.push('\n'); }
            first = false;

            match val {
                // Empty object
                Value::Object(inner) if inner.is_empty() => {
                    self.write_indent(depth, buf);
                    self.write_key(key, buf);
                    buf.push_str(": {}");
                }
                // Empty array
                Value::Array(arr) if arr.is_empty() => {
                    self.write_indent(depth, buf);
                    self.write_key(key, buf);
                    buf.push_str(": []");
                }
                // Primitive value
                v if v.is_primitive() => {
                    self.write_indent(depth, buf);
                    self.write_key(key, buf);
                    buf.push_str(": ");
                    self.write_primitive(v, buf);
                }
                // Array
                Value::Array(arr) => {
                    self.encode_array_value(Some(key), arr, depth, buf);
                }
                // Nested object
                Value::Object(inner) => {
                    self.write_indent(depth, buf);
                    self.write_key(key, buf);
                    buf.push(':');
                    buf.push('\n');
                    self.encode_object(inner, depth + 1, buf);
                }
                _ => {}
            }
        }
    }

    fn encode_array_value(&self, key: Option<&str>, arr: &[Value], depth: usize, buf: &mut String) {
        let len = arr.len();

        if len == 0 {
            self.write_indent(depth, buf);
            if let Some(k) = key {
                self.write_key(k, buf);
                buf.push_str(": []");
            } else {
                buf.push_str("[]");
            }
            return;
        }

        // Array of primitives → inline
        if arr.iter().all(|v| v.is_primitive()) {
            self.write_indent(depth, buf);
            if let Some(k) = key {
                self.write_key(k, buf);
            }
            buf.push('[');
            buf.push_str(&len.to_string());
            buf.push_str("]: ");
            for (i, v) in arr.iter().enumerate() {
                if i > 0 {
                    buf.push(self.delimiter);
                    buf.push(' ');
                }
                self.write_primitive(v, buf);
            }
            return;
        }

        // Array of objects with uniform keys → tabular
        if let Some(fields) = self.extract_tabular_fields(arr) {
            self.write_indent(depth, buf);
            if let Some(k) = key {
                self.write_key(k, buf);
            }
            buf.push('[');
            buf.push_str(&len.to_string());
            buf.push_str("]{");
            for (i, f) in fields.iter().enumerate() {
                if i > 0 { buf.push(self.delimiter); }
                self.write_key(f, buf);
            }
            buf.push_str("}:");
            for item in arr {
                if let Value::Object(obj) = item {
                    buf.push('\n');
                    self.write_indent(depth + 1, buf);
                    for (i, f) in fields.iter().enumerate() {
                        if i > 0 {
                            buf.push(self.delimiter);
                            buf.push(' ');
                        }
                        if let Some(v) = obj.get(f) {
                            self.write_primitive(v, buf);
                        } else {
                            buf.push_str("null");
                        }
                    }
                }
            }
            return;
        }

        // Mixed / expanded array
        self.write_indent(depth, buf);
        if let Some(k) = key {
            self.write_key(k, buf);
        }
        buf.push('[');
        buf.push_str(&len.to_string());
        buf.push_str("]:");
        for item in arr {
            buf.push('\n');
            match item {
                v if v.is_primitive() => {
                    self.write_indent(depth + 1, buf);
                    buf.push_str("- ");
                    self.write_primitive(v, buf);
                }
                Value::Object(obj) if obj.is_empty() => {
                    self.write_indent(depth + 1, buf);
                    buf.push_str("- {}");
                }
                Value::Object(obj) => {
                    self.encode_object_as_list_item(obj, depth + 1, buf);
                }
                Value::Array(inner) if inner.is_empty() => {
                    self.write_indent(depth + 1, buf);
                    buf.push_str("- []");
                }
                Value::Array(inner) if inner.iter().all(|v| v.is_primitive()) => {
                    self.write_indent(depth + 1, buf);
                    buf.push_str("- [");
                    buf.push_str(&inner.len().to_string());
                    buf.push_str("]: ");
                    for (i, v) in inner.iter().enumerate() {
                        if i > 0 {
                            buf.push(self.delimiter);
                            buf.push(' ');
                        }
                        self.write_primitive(v, buf);
                    }
                }
                _ => {}
            }
        }
    }

    fn encode_object_as_list_item(&self, obj: &IndexMap<String, Value>, depth: usize, buf: &mut String) {
        if obj.is_empty() {
            self.write_indent(depth, buf);
            buf.push_str("- ");
            return;
        }

        let mut iter = obj.iter();
        let (first_key, first_val) = iter.next().unwrap();

        self.write_indent(depth, buf);
        buf.push_str("- ");
        self.write_key(first_key, buf);

        match first_val {
            v if v.is_primitive() => {
                buf.push_str(": ");
                self.write_primitive(v, buf);
            }
            Value::Array(arr) if arr.is_empty() => {
                buf.push_str(": []");
            }
            Value::Array(arr) if arr.iter().all(|v| v.is_primitive()) => {
                buf.push('[');
                buf.push_str(&arr.len().to_string());
                buf.push_str("]: ");
                for (i, v) in arr.iter().enumerate() {
                    if i > 0 {
                        buf.push(self.delimiter);
                        buf.push(' ');
                    }
                    self.write_primitive(v, buf);
                }
            }
            Value::Object(inner) if inner.is_empty() => {
                buf.push_str(": {}");
            }
            Value::Object(inner) => {
                buf.push(':');
                buf.push('\n');
                self.encode_object(inner, depth + 2, buf);
            }
            _ => {
                buf.push(':');
            }
        }

        // Continuation fields
        for (key, val) in iter {
            buf.push('\n');
            match val {
                v if v.is_primitive() => {
                    self.write_indent(depth + 1, buf);
                    self.write_key(key, buf);
                    buf.push_str(": ");
                    self.write_primitive(v, buf);
                }
                Value::Array(arr) if arr.is_empty() => {
                    self.write_indent(depth + 1, buf);
                    self.write_key(key, buf);
                    buf.push_str(": []");
                }
                Value::Array(arr) => {
                    self.encode_array_value(Some(key), arr, depth + 1, buf);
                }
                Value::Object(inner) if inner.is_empty() => {
                    self.write_indent(depth + 1, buf);
                    self.write_key(key, buf);
                    buf.push_str(": {}");
                }
                Value::Object(inner) => {
                    self.write_indent(depth + 1, buf);
                    self.write_key(key, buf);
                    buf.push(':');
                    buf.push('\n');
                    self.encode_object(inner, depth + 2, buf);
                }
                _ => {}
            }
        }
    }

    fn extract_tabular_fields(&self, arr: &[Value]) -> Option<Vec<String>> {
        if arr.is_empty() { return None; }

        let first = match &arr[0] {
            Value::Object(obj) if !obj.is_empty() => obj,
            _ => return None,
        };

        let keys: Vec<String> = first.keys().cloned().collect();

        // All values must be primitive
        for v in first.values() {
            if !v.is_primitive() { return None; }
        }

        for item in &arr[1..] {
            match item {
                Value::Object(obj) => {
                    if obj.len() != keys.len() { return None; }
                    for k in &keys {
                        match obj.get(k) {
                            Some(v) if v.is_primitive() => {}
                            _ => return None,
                        }
                    }
                }
                _ => return None,
            }
        }

        Some(keys)
    }

    // --- Primitive writing ---

    fn write_primitive(&self, value: &Value, buf: &mut String) {
        match value {
            Value::Null => buf.push_str("null"),
            Value::Bool(true) => buf.push_str("true"),
            Value::Bool(false) => buf.push_str("false"),
            Value::Integer(n) => buf.push_str(&n.to_string()),
            Value::Float(n) => {
                let s = format!("{}", n);
                buf.push_str(&s);
            }
            Value::String(s) => self.write_string(s, buf),
            _ => {}
        }
    }

    fn write_string(&self, s: &str, buf: &mut String) {
        if self.is_safe_unquoted(s) {
            buf.push_str(s);
        } else {
            buf.push('"');
            self.escape_string(s, buf);
            buf.push('"');
        }
    }

    fn write_key(&self, key: &str, buf: &mut String) {
        if self.is_valid_unquoted_key(key) {
            buf.push_str(key);
        } else {
            buf.push('"');
            self.escape_string(key, buf);
            buf.push('"');
        }
    }

    fn escape_string(&self, s: &str, buf: &mut String) {
        for c in s.chars() {
            match c {
                '\\' => buf.push_str("\\\\"),
                '"' => buf.push_str("\\\""),
                '\n' => buf.push_str("\\n"),
                '\r' => buf.push_str("\\r"),
                '\t' => buf.push_str("\\t"),
                ';' => buf.push_str("\\;"),
                _ => buf.push(c),
            }
        }
    }

    fn is_safe_unquoted(&self, s: &str) -> bool {
        if s.is_empty() { return false; }
        if matches!(s, "true" | "false" | "null") { return false; }
        // Numeric check
        if s.parse::<i64>().is_ok() || s.parse::<f64>().is_ok() { return false; }
        // Delimiter check
        if s.contains(self.delimiter) { return false; }
        // Special chars
        for c in s.chars() {
            if matches!(c, ' ' | '\t' | ':' | '"' | '\\' | ';' | '@' | '#' | '{' | '[' | ']' | '}') {
                return false;
            }
        }
        true
    }

    fn is_valid_unquoted_key(&self, key: &str) -> bool {
        if key.is_empty() { return false; }
        if key.starts_with('#') { return false; }
        for c in key.chars() {
            if matches!(c, ':' | '[' | ']' | '{' | '}' | '"' | '\\' | ' ' | '\t' | ';' | '@' | '#' | ',') {
                return false;
            }
        }
        true
    }

    fn write_indent(&self, depth: usize, buf: &mut String) {
        let spaces = self.indent * depth;
        for _ in 0..spaces {
            buf.push(' ');
        }
    }
}

impl Default for Encoder {
    fn default() -> Self {
        Self::new()
    }
}
